wikimedia/wikipedia-ios
# Allow native article description to update EN Wikipedia wikitext

**Fixes Phabricator ticket:** https://phabricator.wikimedia.org/T257867

### Notes
This PR should fix our broken English article description update flow. It allows us to use our native article description screen to update English Wikipedia, for both new descriptions and existing descriptions. 

The general approach is to move any Wikidata-specific code out of `DescriptionEditViewController.swift`, and instead have it reference a protocol that performs the update specifics depending on which implementation we have passed in. Some other bug fixes and cleanup was added along the way. 

This is best reviewed commit-by-commit. Details are below:

**Commit 1**
Perform some general class renaming and cleanup first. `WikidataDescriptionEditingController` is a Fetcher, but you wouldn't know it from the name. Changed the name to `WikidataFetcher`. Changed to previous `WikidataFetcher` to an extension under `WikidataFetcher+Places.swift`, since it was only used in that one location. I am also removing the reference to `WikidataDescriptionEditingController` out of `MWKDataStore`. It does not use any persistent store within it's calls, so I see no need for it to hang off of that data store object.

**Commit 2**
Commit making tweaks to the JS bridge calls. We need to fetch the current description for pre-populating the textfield on the native screen. Non-EN articles already have this as a part of the wikidata description fetched from the summary endpoint, but EN articles do not. We already have this loaded in the web view so I'm choosing to get it from there rather than making any additional endpoint calls. This also adds some slight cleanup to the ArticleDescriptionSource enum. There's already a .none option, so made sense to lean on that rather than use an optional. This allows me to not deal with the optional later on when presenting the `DescriptionEditViewController`.

**Commit 3**
Tweaks needed to our fetchers and processing. The Short Description API (which we'd be using if it wasn't for oAuth) takes `baserevid` into [account](https://github.com/wikimedia/mobileapps/blob/3a3ac209ddd3ce4e7b676f331c28b4b1afb79520/lib/mwapi.js#L486), so this is pulling the revisionID when fetching the section, to add in later as a `baserevid` when updating the section.

**Commit 4**
This is the bulk of the new logic. Establishes a protocol for updating article descriptions. Copies Wikidata-specific logic out of `DescriptionEditViewController` and into `WikidataDescriptionController`. `ShortDescriptionController` contains the new logic for updating an EN Wikipedia article description. It detects if it already has a Short Description template, and if so, update the description within it. Otherwise prepend a new Short Description template with the new description to the first section. It also uses our JS bridge call in Commit 2 to fetch the article's current description for pre-populating the textfield in `DescriptionEditViewController`.

**Commit 5**
Here we are updating `DescriptionEditViewController` and `ArticleViewController+Editing` to use our new protocol & implementing controllers.

**Commit 6**
Adding some basic tests for `ShortDescriptionController`. Also I noticed when going through the "Learn more" link for EN Wikipedia, the setting to suppress the extra navigation buttons are lost when you navigate to a link within the web view. Here I'm passing this setting along in our `presentOrPush` method, I think it's fairly safe.

**Commit 7**
This fixes a bug I found when updating a non-EN Wikipedia article. Although the article content would reflect the new Wikidata description, the textfield would not pre-populate the latest description when tapping to edit the article description again. This is because the summary endpoint is not attempting to fetch again after performing an update. It seems our [cachePolicy](https://github.com/wikimedia/wikipedia-ios/blob/3fe61d389368499920a175ba8271b68d50536f4b/Wikipedia/Code/ArticleViewController.swift#L434) setting wasn't taking effect because the view state wasn't right (it already switches back to .loaded by this point) and deep within the summary fetching, it totally drops the cachePolicy we have set anyways. This commit fixes these 2 things.

### Test Steps
1. In EN Wikipedia, test adding a new article description and updating an existing article description. Update should go through and you should see it reflected in the article content. Check the wikitext in section 0 (either on Desktop or by choosing to "Edit introduction" in the alert view) - confirm that a {{Short description}} template has been added. Also confirm that the "Learn more" link on the Add/Edit Description modal goes to a web view about Short Descriptions.
2. In any other language Wikipedia, test adding a new article description and updating an existing article description. Update should go through and you should see it reflected in the article content. Check the wikitext in section 0 (either on Desktop or by choosing to "Edit introduction" in the alert view) - confirm that there is NO {{Short description}} template. Go to Wikidata page for that article on desktop, confirm you do see your latest update. Also confirm that the "Learn more" link on the Add/Edit Description modal goes to the FAQ page about Wikidata as it did before.

**Final FYIs**
I have not fixed any abuse filter or captcha handling here. I was able to intermittently reproduce an abuse warning on Test Wiki, but it seemed messed up in our standard editing flow as well, and this PR is already large enough.

Pretty often you will see an error after updating about the article content not being able to refresh in time. I believe this is existing and expected functionality [from this PR](https://github.com/wikimedia/wikipedia-ios/pull/3564).
